examples/3d/custom_render.rs

```diff
@@ -0,0 +1,291 @@
+use bevy::{
+    core::AsBytes,
+    prelude::*,
+    render::{
+        mesh::{INDEX_BUFFER_ASSET_INDEX, VERTEX_BUFFER_ASSET_INDEX},
+        pipeline::{DynamicBinding, PipelineDescriptor, PipelineSpecialization, RenderPipeline},
+        render_graph::{
+            base, AssetRenderResourcesNode, CommandQueue, Node, RenderGraph, ResourceSlots,
+            SystemNode,
+        },
+        renderer::{
+            BufferId, BufferInfo, BufferUsage, RenderContext, RenderResourceBindings,
+            RenderResourceContext, RenderResourceId, RenderResources,
+        },
+        shader::{ShaderStage, ShaderStages},
+    },
+};
+
+/// This example illustrates how to create a custom material and mesh asset and a shader that uses those assets
+fn main() {
+    App::build()
+        .add_default_plugins()
+        .add_asset::<MyMesh>()
+        .add_asset::<MyMaterial>()
+        .add_startup_system(setup.system())
+        .run();
+}
+
+#[derive(Default)]
+pub struct MyMesh;
+
+#[derive(Default)]
+pub struct MyMeshNode {
+    command_queue: CommandQueue,
+}
+
+impl Node for MyMeshNode {
+    fn update(
+        &mut self,
+        _world: &World,
+        _resources: &Resources,
+        render_context: &mut dyn RenderContext,
+        _input: &ResourceSlots,
+        _output: &mut ResourceSlots,
+    ) {
+        self.command_queue.execute(render_context);
+    }
+}
+
+impl SystemNode for MyMeshNode {
+    fn get_system(&self, commands: &mut Commands) -> Box<dyn System> {
+        let system = my_mesh_node_system.system();
+        commands.insert_local_resource(
+            system.id(),
+            MyMeshNodeState {
+                command_queue: self.command_queue.clone(),
+                ..Default::default()
+            },
+        );
+
+        system
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+struct MyMeshBuffers {
+    vertex: BufferId,
+    index: BufferId,
+    vertex_staging: BufferId,
+    index_staging: BufferId,
+}
+
+impl MyMeshBuffers {
+    pub fn new(
+        render_resource_context: &dyn RenderResourceContext,
+        vertex: BufferInfo,
+        index: BufferInfo,
+    ) -> Self {
+        Self {
+            vertex: render_resource_context.create_buffer(BufferInfo {
+                buffer_usage: vertex.buffer_usage | BufferUsage::COPY_DST,
+                mapped_at_creation: false,
+                ..vertex
+            }),
+            index: render_resource_context.create_buffer(BufferInfo {
+                buffer_usage: index.buffer_usage | BufferUsage::COPY_DST,
+                mapped_at_creation: false,
+                ..index
+            }),
+            vertex_staging: render_resource_context.create_buffer(BufferInfo {
```

Comment:
Just as a unrelated note, we likely should be using `wgpu::queue.write_buffer` for some stuff as it removes the need for a staging buffer.

Reply:
That's really helpful, thanks.